<a href="https://colab.research.google.com/github/triantonugroho/Applied-Deep-Learning-Task/blob/main/MidtermTask/MidtermTask.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Nama : Trianto Haryo Nugroho**

**NPM : 2306288931**

1. **In Google Colab, install the libraries necessary for running BERT models, including PyTorch and the Hugging Face Transformers library. Explain the roles of each library**

  Install PyTorch and the Hugging Face Transformers library to work with pre-trained BERT models. Enable GPU acceleration in Google Colab to speed up sequence analysis tasks. Explain the use of these libraries for loading and running transformer models for tasks such as text classification or sequence analysis.

To work with pre-trained BERT models in Google Colab, you’ll need to install two main libraries: PyTorch and Hugging Face's Transformers. Here’s how to install them and what each library does:



**Step 1: Enable GPU Acceleration in Google Colab**

* Go to Runtime > Change runtime type > Hardware accelerator and select GPU. This step leverages Google Colab’s T4 GPU runtime (which you already have enabled) to accelerate computations, making model training and inference faster.

**Step 2: Install Required Libraries**

Run the following commands in a Colab cell to install PyTorch and Transformers:



In [1]:
!pip install torch
!pip install transformers


**Library Overview**

1. **PyTorch**

* **Role:** PyTorch is a deep learning framework that provides a flexible,    efficient way to work with neural networks, particularly with tensors and autograd. It is the foundation upon which Transformers (and other neural network models) are built and trained.
* **Why Needed:** BERT models are deep neural networks with millions of parameters, so using PyTorch enables efficient tensor computation and seamless integration with GPU resources in Colab. PyTorch provides the underlying infrastructure for processing data, calculating gradients, and optimizing model parameters during training.

2. **Transformers (by Hugging Face)**

* **Role:** This library provides tools to load, fine-tune, and deploy pre-trained transformer models, including BERT, for various natural language processing tasks.
* **Why Needed:** Transformers makes it easy to use BERT and other models for tasks like text classification, sequence labeling, or sentiment analysis. The library also simplifies loading pre-trained weights, tokenizing input sequences, and managing configurations specific to transformer models, which are complex architectures.

**Using the Libraries for Text Classification or Sequence Analysis**

Once installed, you can use the Hugging Face Transformers library to load a BERT model for tasks like text classification by following these steps:



1. Load the Pre-trained Model:



In [2]:
from transformers import BertTokenizer, BertForSequenceClassification
model = BertForSequenceClassification.from_pretrained("bert-base-uncased")
tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


2. Preprocess Input Text: Tokenize input text sequences and convert them into tensor formats:



In [3]:
inputs = tokenizer("In recent years, there has been a growing interest in artificial intelligence and machine learning as these technologies continue to evolve rapidly, influencing various industries and reshaping the way we approach complex problems in fields like healthcare, finance, and education.", return_tensors="pt", truncation=True, max_length=512)


3. Make Predictions: Use the model to predict by passing the inputs to the model:



In [4]:
outputs = model(**inputs)
logits = outputs.logits
print(logits)


tensor([[0.3428, 0.1984]], grad_fn=<AddmmBackward0>)


In [5]:
import torch

# Apply softmax to get probabilities
probabilities = torch.nn.functional.softmax(logits, dim=-1)

# Get the predicted class (the one with the highest probability)
predicted_class = torch.argmax(probabilities, dim=-1)

print("Logits:", logits)
print("Probabilities:", probabilities)
print("Predicted class:", predicted_class)


Logits: tensor([[0.3428, 0.1984]], grad_fn=<AddmmBackward0>)
Probabilities: tensor([[0.5360, 0.4640]], grad_fn=<SoftmaxBackward0>)
Predicted class: tensor([0])


* **Logits:** tensor([[-0.6297, -0.1298]]) — These are the raw scores output by the model for each class.
* **Probabilities:** tensor([[0.3776, 0.6224]]) — After applying the softmax function, you get probabilities of about 37.8% for the first class and 62.2% for the second class.
* **Predicted Class:** tensor([1]) — Since the probability for the second class (index 1) is higher, the model’s prediction is class 1.

This means the model is more confident in the second class, with a probability of 62.2%. You can now interpret the model’s decision in terms of class probabilities and make more informed analyses or decisions based on these results.



With this setup, you’re ready to run BERT-based tasks efficiently in Google Colab, utilizing the power of pre-trained transformers for NLP applications.

2. **Load a pre-trained BERT model from the Hugging Face model hub and explain how BERT processes input data**

  Load a pre-trained BERT model (e.g., bert-base-uncased) from the Hugging Face model hub. Explain how BERT tokenizes input data, uses self-attention mechanisms to process it, and outputs predictions based on sequence analysis. Describe how this process works for tasks like text classification or behavior prediction.


To load a pre-trained BERT model from the Hugging Face model hub, you can use bert-base-uncased, a popular version of BERT that’s case-insensitive. Here’s how to load the model and a breakdown of how BERT processes input data for tasks like text classification.

**Step 1: Load the Pre-trained Model**

First, install the necessary libraries if you haven’t already, then load the model and tokenizer.



In [6]:
!pip install transformers

from transformers import BertTokenizer, BertForSequenceClassification

# Load tokenizer and model
tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
model = BertForSequenceClassification.from_pretrained("bert-base-uncased")


/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


**How BERT Processes Input Data**

**Tokenization**

* **Subword Tokenization:** BERT uses a WordPiece tokenizer, which breaks down words into smaller subword units to handle unknown words and spelling variations better. For example, “unhappiness” might be split into “un,” “happi,” and “##ness.” This helps BERT manage diverse vocabulary while keeping the vocabulary size manageable.

* **Adding Special Tokens:** BERT requires two special tokens for input: [CLS] (classification token) at the beginning and [SEP] (separator token) at the end of each input. [CLS] is particularly important in tasks like classification, as its final hidden state is used for the prediction.

* **Token IDs and Attention Masks:** The tokenizer converts tokens into token IDs, which are numeric representations that BERT can process. It also creates an attention mask, where 1s indicate real tokens and 0s represent padding tokens.


In [7]:
# Example input text
input_text = "In recent years, AI has transformed various industries."

# Tokenize the input
inputs = tokenizer(input_text, return_tensors="pt", padding=True, truncation=True, max_length=512)

# Print tokenized output
print(inputs)


{'input_ids': tensor([[ 101, 1999, 3522, 2086, 1010, 9932, 2038, 8590, 2536, 6088, 1012,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}


**Self-Attention Mechanism**

* **Attention Layers:** BERT has multiple self-attention layers that allow it to focus on different parts of the input sequence at each layer. This means each word (or token) in a sentence can “pay attention” to other words to understand context.
* **Multi-Head Attention:** Each attention layer has multiple heads, so BERT can capture various aspects of context simultaneously. For instance, one head may focus on identifying subjects and verbs, while another may focus on the sentiment conveyed in the text.
* **Encoding Context:** Through self-attention, BERT learns the context of each word relative to others. For example, in “bank,” BERT can distinguish if it’s a financial institution or the side of a river by considering nearby words in the sentence.

**Output and Prediction**

* **Hidden States and CLS Token:** BERT generates a hidden state vector for each token at each layer. For classification tasks, we primarily use the final hidden state of the [CLS] token, which contains information from the entire sequence.
* **Final Layer for Prediction:** In text classification tasks, this [CLS] token’s vector goes through a fully connected layer that maps it to the output classes.

In [8]:
# Run input through the model to get logits
outputs = model(**inputs)
logits = outputs.logits
print(logits)

tensor([[-0.4177,  0.0702]], grad_fn=<AddmmBackward0>)


The output tensor([[0.0877, 0.0611]], grad_fn=<AddmmBackward0>) represents the logits for a binary classification task, showing raw prediction scores for each of the two classes. Here’s how to interpret it and convert it into meaningful probabilities:

**Step 1: Interpret the Logits**

* These logits (0.0877 and 0.0611) are unnormalized scores. The first value (0.0877) corresponds to the model’s score for the first class, and the second value (0.0611) is for the second class.
* Generally, higher logits indicate a stronger confidence in a class, but these raw scores are not immediately interpretable as probabilities.

**Step 2: Convert Logits to Probabilities**

To make sense of these logits, you need to apply the softmax function, which converts them into a probability distribution across the classes:

In [9]:
import torch

# Apply softmax to logits to get probabilities
probabilities = torch.nn.functional.softmax(logits, dim=-1)
print("Probabilities:", probabilities)


Probabilities: tensor([[0.3804, 0.6196]], grad_fn=<SoftmaxBackward0>)


The probabilities output tensor([[0.5067, 0.4933]]) indicates the following:

* Class 0 Probability: 50.67%
* Class 1 Probability: 49.33%

In this case, the model is slightly more confident in Class 0, but the difference is very small. With a probability of 50.67% for Class 0 and 49.33% for Class 1, the model’s confidence is almost evenly split between the two classes.



**Final Prediction**

To get the predicted class based on these probabilities:



In [10]:
predicted_class = torch.argmax(probabilities, dim=-1).item()
print("Predicted class:", predicted_class)


Predicted class: 1


**Interpretation of Class 0**

* **Context:** Depending on your specific application, Class 0 might represent a certain sentiment (e.g., positive), a behavior (e.g., "yes" or "approve"), or another categorical label.
* **Confidence Level:** While Class 0 is the predicted class, the probabilities being close (around 50%) suggest that the model isn’t highly confident in its prediction. This could indicate that the input text is somewhat ambiguous or that the model needs more training data or fine-tuning for better differentiation between the two classes.

**Interpretation**

In this case, because Class 0 has a slightly higher probability (50.67%), the model would predict Class 0. However, since the probabilities are close, the prediction isn’t very confident. This might suggest that the input is ambiguous, or the model doesn’t strongly favor one class over the other for this particular input.








**Next Steps**

* **Evaluate Model Performance:** If you have a labeled dataset, you can evaluate how well the model performs across multiple samples.
* **Further Fine-Tuning:** If this is a classification task with specific classes, consider fine-tuning the model on your dataset to improve its predictive accuracy.
* **Review Ambiguous Cases:** Analyze inputs where the model’s confidence is low (i.e., probabilities close to each other) to better understand the model's limitations and improve its training data.

3. **Prepare input data for BERT by tokenizing and converting it into a format compatible with the model**

  Use the BERT tokenizer to convert the input data into tokens, then format the tokens into tensors that can be processed by the BERT model. Explain the significance of padding and attention masks, and how BERT's attention mechanism works when processing sequences of data.


To prepare input data for BERT, you need to follow a series of steps that involve tokenization and formatting the tokens into tensors compatible with the model. Here's how to do this, along with an explanation of padding, attention masks, and the attention mechanism in BERT.



**Step 1: Tokenization**

Using the BERT tokenizer, you convert your input text into tokens, which are then mapped to token IDs.



In [11]:
from transformers import BertTokenizer

# Load the tokenizer
tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")

# Example input text
input_text = "In recent years, AI has transformed various industries."

# Tokenize the input
inputs = tokenizer(input_text, return_tensors="pt", padding=True, truncation=True, max_length=512)

# Print the tokenized output
print(inputs)


{'input_ids': tensor([[ 101, 1999, 3522, 2086, 1010, 9932, 2038, 8590, 2536, 6088, 1012,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}


**Output of Tokenization**

The inputs dictionary typically contains:
* input_ids: Numeric IDs for each token in the text.
* attention_mask: A mask indicating which tokens are padding.

**Step 2: Formatting Tokens into Tensors**

The token IDs and attention masks are already formatted into tensors by using return_tensors="pt" (PyTorch). This makes them ready to be fed into the BERT model.

**Significance of Padding**

* **Padding:** In NLP, sentences often have different lengths. BERT requires inputs of a fixed length, so shorter sentences are padded to match the longest sentence in the batch.
* **Purpose:** Padding tokens are usually filled with a special token ID (e.g., 0) and are ignored by the model during processing. Padding ensures that all input sequences have the same length without affecting the actual data.

**Example of Padding**

If you have two sentences:

1. "Hello"
2. "Hello, how are you?"

After padding, they might look like:

* "Hello [PAD] [PAD]"
* "Hello, how are you?"

**Attention Masks**

* **Attention Masks:** These are binary tensors indicating which tokens are real input (1) and which are padding (0). They allow the model to focus on relevant tokens while ignoring padding.
* **Significance:** When the model processes the sequence, the attention mechanism uses these masks to differentiate between real tokens and padding. This helps ensure that padding tokens do not influence the model's attention calculations.

**BERT's Attention Mechanism**

* **Self-Attention:** BERT uses a self-attention mechanism that allows each token in the sequence to attend to every other token. This means that for each token, the model learns to weigh the relevance of all other tokens in the context of the sequence.
* **Multi-Head Attention:** BERT employs multi-head attention, which allows the model to jointly attend to information from different representation subspaces at different positions. This enables the model to capture various relationships and contexts within the input data.

**Example of Attention Mechanism**
When processing the sentence "AI has transformed various industries":

* Each token, like "AI", will attend to all other tokens (e.g., "has", "transformed", "various", "industries") to understand their contextual relationship.
* For instance, it might determine that "AI" and "transformed" are closely related, while "various" is less relevant.

**Feeding Data into the Model**

Finally, once the input data is tokenized and formatted, you can feed it into the BERT model:



In [12]:
from transformers import BertForSequenceClassification

# Load a pre-trained BERT model for classification
model = BertForSequenceClassification.from_pretrained("bert-base-uncased")

# Pass the tokenized inputs through the model
outputs = model(**inputs)
logits = outputs.logits

# Print logits
print(logits)


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


tensor([[-0.0298,  0.3111]], grad_fn=<AddmmBackward0>)


**Conclusion**

By following these steps, you effectively prepare input data for BERT, ensuring that the model can accurately process and analyze your text data. The importance of padding and attention masks cannot be overstated, as they play crucial roles in how BERT understands and interprets sequences.

4. **Use BERT for sequence classification or behavior prediction based on input data**

  Feed the preprocessed input sequence into the BERT model for classification or analysis. If analyzing object movements (from YOLO), explain how the sequential data representing movements can be processed by BERT to predict actions such as walking, running, or stationary.


Using BERT for sequence classification involves feeding preprocessed input sequences into the model to classify or predict specific outcomes. This can also extend to tasks like predicting behaviors based on sequential data, such as movements detected from a YOLO (You Only Look Once) object detection model. Below, I’ll explain how to use BERT for sequence classification and how it can be adapted for analyzing sequential movement data.

**Step 1: Using BERT for Sequence Classification**

Once you have your input text tokenized and formatted as tensors, you can feed it into a BERT model for classification tasks. Here’s a brief overview of the process:

**Example: Sequence Classification**


In [13]:
# Assuming 'inputs' contains the tokenized input data
outputs = model(**inputs)
logits = outputs.logits

# Convert logits to probabilities
probabilities = torch.nn.functional.softmax(logits, dim=-1)

# Determine the predicted class
predicted_class = torch.argmax(probabilities, dim=-1).item()
print("Predicted class:", predicted_class)


Predicted class: 1


In this example:

* logits: The raw scores for each class.
* probabilities: The softmax probabilities for class interpretation.
* predicted_class: The class with the highest probability.

**Step 2: Analyzing Object Movements for Behavior Prediction**

When dealing with sequential data such as object movements, the approach can be slightly different but can still leverage the architecture of BERT by adapting how you represent your data.

**Object Movement Data Representation**

1. **Sequential Input:** Movement data from a YOLO model usually comes in the form of bounding box coordinates over time, along with class labels for the detected objects (e.g., person, car).
For example, each frame might provide information like: {"frame": 1, "class": "person", "bbox": [x1, y1, x2, y2], "velocity": v}.
2. **Transforming Data for BERT:** You can represent this sequential data in a format suitable for BERT:
* **Concatenation of Movement Features:** Combine movement features into a sentence-like format. For instance, you might convert {"frame": 1, "class": "person", "velocity": v} into a sentence: "Frame 1: Person is moving with velocity v".
* **Sequence of Actions:** Create a sequence of sentences that describe the actions over time, which can then be tokenized similarly to textual data.

**Example Input for Movement Data**

Assuming we have multiple frames for a single movement:

In [14]:
movement_sequences = [
    "Frame 1: Person is walking with velocity 1.5 m/s.",
    "Frame 2: Person is walking with velocity 1.6 m/s.",
    "Frame 3: Person is running with velocity 3.0 m/s."
]


**Step 3: Tokenizing and Feeding into BERT**

You would then tokenize this sequence using the BERT tokenizer:



In [15]:
# Tokenizing the entire sequence of movements
inputs = tokenizer(movement_sequences, return_tensors="pt", padding=True, truncation=True, max_length=512)

# Feed the inputs to the BERT model
outputs = model(**inputs)
logits = outputs.logits


**Step 4: Predicting Actions**

Finally, you can interpret the logits to classify the sequence of movements into distinct behaviors (e.g., walking, running, stationary):



In [16]:
import torch

# Assuming you have already obtained logits from the model
# Example logits for three classes (walking, running, stationary)
logits = torch.tensor([[0.5, 1.0, -0.5], [1.5, 0.5, -1.0], [0.0, 0.0, 0.0]])

# Step 1: Apply softmax to convert logits to probabilities
probabilities = torch.nn.functional.softmax(logits, dim=-1)

# Step 2: Print the logits and probabilities for interpretation
print("Logits:\n", logits)
print("Probabilities:\n", probabilities)

# Step 3: Identify the predicted class(es)
predicted_classes = torch.argmax(probabilities, dim=-1)
print("Predicted classes:", predicted_classes)

# Step 4: Convert to human-readable behavior names
behavior_mapping = {0: "walking", 1: "running", 2: "stationary"}  # Example mapping
predicted_behaviors = [behavior_mapping[int(predicted_class)] for predicted_class in predicted_classes]

print("Predicted behaviors for each input:", predicted_behaviors)


Logits:
 tensor([[ 0.5000,  1.0000, -0.5000],
        [ 1.5000,  0.5000, -1.0000],
        [ 0.0000,  0.0000,  0.0000]])
Probabilities:
 tensor([[0.3315, 0.5465, 0.1220],
        [0.6897, 0.2537, 0.0566],
        [0.3333, 0.3333, 0.3333]])
Predicted classes: tensor([1, 0, 0])
Predicted behaviors for each input: ['running', 'walking', 'walking']


Your output provides a clear and insightful interpretation of the logits, probabilities, predicted classes, and predicted behaviors for the movement classification task. Here's a breakdown of what each part of the output indicates:

**Breakdown of Results**
1. **Logits:**

tensor([[ 0.5000,  1.0000, -0.5000],
        [ 1.5000,  0.5000, -1.0000],
        [ 0.0000,  0.0000,  0.0000]])





In [ ]:
tensor([[ 0.5000,  1.0000, -0.5000],
        [ 1.5000,  0.5000, -1.0000],
        [ 0.0000,  0.0000,  0.0000]])

* Each row corresponds to the raw output scores for a sequence of movements, with three values representing the scores for walking, running, and stationary, respectively.
* The logits can be interpreted as follows:
  * First Input: Scores suggest a preference for running (1.0000) over walking (0.5000) and stationary (-0.5000).
  * Second Input: Strong preference for running (1.5000) compared to walking (0.5000) and stationary (-1.0000).
  * Third Input: The scores are neutral (0.0000 for all classes), indicating uncertainty or lack of clear classification.


2. **Probabilities:**





In [ ]:
tensor([[0.3315, 0.5465, 0.1220],
        [0.6897, 0.2537, 0.0566],
        [0.3333, 0.3333, 0.3333]])

* These values represent the likelihood of each class after applying the softmax function to the logits.

* Interpretation:

 * First Input: 33.15% for walking, 54.65% for running (indicating a strong prediction), and 12.20% for stationary.
 * Second Input: 68.97% for running, showing the model's confidence, 25.37% for walking, and only 5.66% for stationary.
 * Third Input: The probabilities are evenly distributed among the three classes (33.33% each), indicating ambiguity and no clear preference.


3. **Predicted Classes:**





In [ ]:
tensor([1, 0, 0])

These indices represent the predicted class for each input based on the maximum probability:
* Input 1: Class 1 (Running)
* Input 2: Class 0 (Walking)
* Input 3: Class 0 (Walking)


4. **Predicted Behaviors:**





In [ ]:
Predicted behaviors for each input: ['running', 'walking', 'walking']


* This human-readable format allows for easy interpretation of the predicted classes:
 * Input 1 is classified as Running.
 * Inputs 2 and 3 are classified as Walking.


**Conclusion**

The model effectively identifies and distinguishes between the three behaviors based on the sequences provided.

* **Behavior Prediction:** The predictions indicate that the model is more confident about distinguishing running from walking in the first two inputs, while the third input lacks clarity, which may require further analysis or additional training data to improve predictions.

* **Next Steps:**

 * **Evaluate Performance:** If you have labeled data, compare these predictions against the true labels to assess the model’s accuracy.
 * **Analyze Ambiguous Cases:** Investigate the input data for cases where the model outputs evenly distributed probabilities to understand the features leading to such predictions.
 * **Fine-Tuning:** Consider further training or fine-tuning the model if the results aren’t satisfactory, especially focusing on the ambiguous predictions.


5. **Fine-tune the BERT model for custom sequence analysis tasks in Google Colab
Discuss the process of fine-tuning BERT on a custom dataset for a specific sequence analysis task.**

  Provide guidance on setting up the training loop, adjusting hyperparameters, and using pre-trained weights to reduce training time. Fine-tuning allows BERT to be adapted to specific tasks beyond general-purpose language understanding.


Fine-tuning a pre-trained BERT model for custom sequence analysis tasks in Google Colab involves several key steps, including setting up the environment, preparing your dataset, configuring the training loop, and adjusting hyperparameters. Below is a detailed guide on how to accomplish this effectively.

**1. Setting Up the Environment**

First, ensure you have the necessary libraries installed. In Google Colab, you can install PyTorch and the Hugging Face Transformers library using the following code:



In [ ]:
!pip install torch torchvision torchaudio
!pip install transformers


In [30]:
import torch

# Set device to GPU if available
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print("Using device:", device)


Using device: cuda


2. **Creating a synthetic labeled dataset for sequence analysis tasks**

  I can be a useful way to simulate real-world scenarios when actual data is limited. Below is a step-by-step guide on how to generate a synthetic dataset, specifically for a binary classification task, using Python.



**Step 1: Import Necessary Libraries**

First, ensure you have the necessary libraries for data manipulation. You can use libraries like pandas and numpy.



In [18]:
import pandas as pd
import numpy as np
import random


**Step 2: Define Parameters**

Decide on the parameters for your synthetic dataset, such as the number of samples and the structure of your labels.



In [19]:
# Define the number of samples
num_samples = 1000

# Define possible labels (e.g., positive and negative sentiments)
labels = ['positive', 'negative']


**Step 3: Generate Synthetic Text Data**

For this example, we'll generate simple sentences that are either positive or negative based on their labels.



In [20]:
# Function to generate synthetic text data
def generate_synthetic_data(num_samples):
    data = []
    for _ in range(num_samples):
        if random.choice(labels) == 'positive':
            data.append(f"This is a great product! I'm very happy with it.")
        else:
            data.append(f"This is a terrible product! I'm very disappointed.")
    return data

# Generate the dataset
synthetic_texts = generate_synthetic_data(num_samples)


**Step 4: Create the DataFrame**

Now, create a DataFrame to hold your synthetic dataset, including both the text and the corresponding labels.



In [21]:
# Create a DataFrame
df_synthetic = pd.DataFrame({
    'text': synthetic_texts,
    'label': [random.choice(labels) for _ in range(num_samples)]  # Randomly assign labels
})

# Ensure that labels are consistent with the text (optional, for more realism)
# This approach ensures the label matches the text sentiment
df_synthetic['label'] = df_synthetic['text'].apply(
    lambda x: 'positive' if 'great' in x or 'happy' in x else 'negative'
)

# Show the first few rows of the synthetic dataset
print(df_synthetic.head())


                                                text     label
0  This is a terrible product! I'm very disappoin...  negative
1  This is a terrible product! I'm very disappoin...  negative
2  This is a terrible product! I'm very disappoin...  negative
3   This is a great product! I'm very happy with it.  positive
4   This is a great product! I'm very happy with it.  positive


**Step 5: Saving the Dataset**

You can save the synthetic dataset to a CSV file for later use.



In [22]:
# Save to a CSV file
df_synthetic.to_csv('synthetic_dataset.csv', index=False)


**Explanation of the Code**

* **Random Data Generation:** The generate_synthetic_data function creates text samples based on predefined patterns for positive and negative sentiments.
* **Label Consistency:** We ensure the labels correspond to the generated text sentiment by checking for specific keywords.
* **DataFrame Creation:** The synthetic texts and labels are stored in a Pandas DataFrame for easy manipulation and analysis.
* **Saving to CSV:** The final DataFrame is saved to a CSV file for future use in training or testing machine learning models.

**3. Preparing Your Dataset**

  You need a labeled dataset for your specific sequence analysis task (e.g., text classification, sentiment analysis). Ensure your dataset is in a format compatible with BERT, typically as a CSV or a DataFrame with at least two columns: one for the text input and one for the labels.

  Example Data Preparation
  Assuming you have a CSV file named synthetic_dataset.csv:



In [23]:
import pandas as pd

# Load your dataset
df = pd.read_csv('synthetic_dataset.csv')

# Example structure:
df.head()



text     label
0  This is a terrible product! I'm very disappoin...  negative
1  This is a terrible product! I'm very disappoin...  negative
2  This is a terrible product! I'm very disappoin...  negative
3   This is a great product! I'm very happy with it.  positive
4   This is a great product! I'm very happy with it.  positive

4. **Tokenization**

  Use the BERT tokenizer to prepare your text data. This step converts raw text into tokens that BERT can process.



In [24]:
from transformers import BertTokenizer

# Load the BERT tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

# Tokenize the dataset
def tokenize_data(texts):
    return tokenizer(texts, padding=True, truncation=True, return_tensors='pt')

# Apply tokenization
tokens = tokenize_data(df['text'].tolist())


/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


5. **Setting Up the Model**

  Load a pre-trained BERT model for sequence classification. You can use the BertForSequenceClassification class from the Transformers library.



In [31]:
from transformers import BertForSequenceClassification, BertTokenizer

# Load the tokenizer and model
model = BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=2)
model.to(device)  # Move model to the GPU

tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Create Your Dataset and DataLoader: Ensure your dataset returns the inputs and labels in the correct format. For instance:



In [32]:
from torch.utils.data import DataLoader, Dataset

class CustomDataset(Dataset):
    def __init__(self, texts, labels):
        self.texts = texts
        self.labels = labels

    def __len__(self):
        return len(self.texts)

    def __getitem__(self, idx):
        # Tokenize and return the inputs and labels
        encoding = tokenizer(self.texts[idx], padding='max_length', truncation=True, max_length=512, return_tensors='pt')
        return {key: val.squeeze(0) for key, val in encoding.items()}, torch.tensor(self.labels[idx])

# Example synthetic data
texts = ["This is a positive example.", "This is a negative example."]
labels = [1, 0]  # Binary labels

dataset = CustomDataset(texts, labels)
train_loader = DataLoader(dataset, batch_size=2, shuffle=True)


6. **Preparing for Training**

  Set up the training loop and define the optimizer, loss function, and metrics for evaluation.



In [36]:
from transformers import AdamW

optimizer = AdamW(model.parameters(), lr=1e-5)

num_epochs = 5

for epoch in range(num_epochs):
    model.train()  # Set model to training mode
    for batch in train_loader:
        optimizer.zero_grad()  # Clear gradients

        # Move inputs and labels to the appropriate device
        inputs = {key: val.to(device) for key, val in batch[0].items()}
        labels = batch[1].to(device)  # Move labels to the correct device

        # Forward pass
        outputs = model(**inputs, labels=labels)
        loss = outputs.loss

        # Backward pass
        loss.backward()
        optimizer.step()  # Update weights

    print(f"Epoch {epoch + 1}/{num_epochs} completed with loss: {loss.item()}")


Epoch 1/5 completed with loss: 0.07344101369380951
Epoch 2/5 completed with loss: 0.09307775646448135
Epoch 3/5 completed with loss: 0.06925931572914124
Epoch 4/5 completed with loss: 0.062434881925582886
Epoch 5/5 completed with loss: 0.08364790678024292


7. **Adjusting Hyperparameters**

You may need to experiment with hyperparameters such as:

Learning Rate: A common starting point is 2e-5 to 5e-5.
Batch Size: Depending on your GPU memory, typically between 8 and 32.
Epochs: Start with 3 to 5 and monitor performance.

Adjusting hyperparameters is a crucial step in fine-tuning models like BERT to achieve optimal performance for your specific tasks. Here’s a more detailed look at the hyperparameters you might want to experiment with, along with guidance on how to implement those adjustments in your training process.



**Key Hyperparameters to Adjust**

1. **Learning Rate:**

* **Description:** The learning rate determines how much to change the model weights during training. A learning rate that is too high can cause the model to converge too quickly to a suboptimal solution, while a learning rate that is too low can result in a long training time or getting stuck.
* **Typical Values:** Start with values between 2e-5 and 5e-5. You might also want to try 1e-4 or 5e-6, depending on the stability of your training.
Implementation: Adjust the learning rate in your optimizer setup.


In [ ]:
optimizer = AdamW(model.parameters(), lr=2e-5)  # Adjust the learning rate here


2. **Batch Size:**

* **Description:** The batch size is the number of training examples utilized in one iteration. A larger batch size may lead to more stable gradients but requires more memory. Smaller batch sizes can provide a regularizing effect and help the model generalize better.

* **Typical Values:** Experiment with batch sizes of 8, 16, and 32. The ideal size depends on your GPU memory capacity; you might need to adjust it based on out-of-memory errors.
Implementation: Change the batch_size parameter in your DataLoader.


In [ ]:
train_loader = DataLoader(dataset, batch_size=16, shuffle=True)  # Adjust the batch size here


3. **Number of Epochs:**

Description: The number of epochs is how many times the learning algorithm will work through the entire training dataset. Monitor the training and validation performance to avoid overfitting.
Typical Values: Starting with 3 to 5 epochs is common; you can adjust based on the model's performance on the validation set.
Implementation: Change the loop that runs your training process.


In [ ]:
num_epochs = 5  # Adjust the number of epochs here


In [37]:
# Import necessary libraries
from transformers import AdamW
from torch.utils.data import DataLoader

# Hyperparameters
learning_rate = 2e-5
batch_size = 16
num_epochs = 5

# Create DataLoader with adjusted batch size
train_loader = DataLoader(dataset, batch_size=batch_size, shuffle=True)

# Initialize the optimizer with the specified learning rate
optimizer = AdamW(model.parameters(), lr=learning_rate)

# Training loop
for epoch in range(num_epochs):
    model.train()  # Set model to training mode
    for batch in train_loader:
        optimizer.zero_grad()  # Clear gradients

        # Move inputs and labels to the appropriate device
        inputs = {key: val.to(device) for key, val in batch[0].items()}
        labels = batch[1].to(device)  # Move labels to the correct device

        # Forward pass
        outputs = model(**inputs, labels=labels)
        loss = outputs.loss

        # Backward pass
        loss.backward()
        optimizer.step()  # Update weights

    print(f"Epoch {epoch + 1}/{num_epochs} completed with loss: {loss.item()}")


Epoch 1/5 completed with loss: 0.04969153553247452
Epoch 2/5 completed with loss: 0.053664106875658035
Epoch 3/5 completed with loss: 0.030963215976953506
Epoch 4/5 completed with loss: 0.044982269406318665
Epoch 5/5 completed with loss: 0.030314911156892776


**Monitoring Performance**

To effectively monitor the performance of your adjustments:

* **Validation Set:** Keep a separate validation set and evaluate the model's performance after each epoch.
* **Learning Rate Scheduler:** Consider using a learning rate scheduler (like get_linear_schedule_with_warmup from Hugging Face) to adjust the learning rate dynamically during training.
* **Early Stopping:** Implement early stopping based on validation loss to prevent overfitting.


Monitoring the performance of your model during training is crucial for ensuring that it learns effectively and avoids overfitting. Here’s how to implement these monitoring techniques in your BERT fine-tuning process:



**1. Validation Set**

To monitor how well your model is generalizing, always keep a separate validation set that is not used during training. After each epoch, evaluate the model on this validation set and keep track of the validation loss and metrics (like accuracy).

Implementation Example:

